From aeadcd4ba6d2a49fb5bcdc71b6a166d5bcc1415c Mon Sep 17 00:00:00 2001
From: Keunhong Park <keunhong@codeaurora.org>
Date: Thu, 16 Aug 2012 17:38:01 -0600
Subject: [PATCH] Add web servlet for multi-master plugin

This shows the status of a server's peers and recent cache evicts.

Change-Id: Iebbacf00dd31f206e1bc1737873957d7b55fd289
---
 html/index.html                                    |   46 ++++++++
 pom.xml                                            |    3 +
 .../gerrit/plugins/multimaster/DataAggregator.java |  120 ++++++++++++++++++++
 .../plugins/multimaster/MultiMasterModule.java     |    2 +
 .../plugins/multimaster/servlet/JsonServlet.java   |   58 ++++++++++
 .../multimaster/servlet/MultiMasterServlet.java    |   78 +++++++++++++
 .../servlet/MultiMasterServletModule.java          |   37 ++++++
 static/css/status.css                              |   42 +++++++
 static/js/status.js                                |   98 ++++++++++++++++
 9 files changed, 484 insertions(+), 0 deletions(-)
 create mode 100644 html/index.html
 create mode 100644 src/main/java/com/google/gerrit/plugins/multimaster/DataAggregator.java
 create mode 100644 src/main/java/com/google/gerrit/plugins/multimaster/servlet/JsonServlet.java
 create mode 100644 src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServlet.java
 create mode 100644 src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServletModule.java
 create mode 100644 static/css/status.css
 create mode 100644 static/js/status.js

diff --git a/html/index.html b/html/index.html
new file mode 100644
index 0000000..7114486
--- /dev/null
+++ b/html/index.html
@@ -0,0 +1,46 @@
+<!doctype html>
+<html>
+<!-- 
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ -->
+<head>
+<title>Gerrit MultiMaster Status</title>
+<link
+	href="http://fonts.googleapis.com/css?family=Source+Sans+Pro:400,600"
+	rel="stylesheet" type="text/css" />
+<link href="../static/css/status.css" rel="stylesheet" type="text/css" />
+<script type="text/javascript"
+	src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
+<script type="text/javascript" src="../static/js/status.js"></script>
+</head>
+<body>
+	<h1>Gerrit MultiMaster Status</h1>
+	<div id="wrapper"></div>
+</body>
+</html>
+1
diff --git a/pom.xml b/pom.xml
index ed68169..e870d77 100644
--- a/pom.xml
+++ b/pom.xml
@@ -38,6 +38,7 @@ limitations under the License.
           <archive>
             <manifestEntries>
               <Gerrit-Module>com.google.gerrit.plugins.multimaster.MultiMasterModule</Gerrit-Module>
+              <Gerrit-HttpModule>com.google.gerrit.plugins.multimaster.servlet.MultiMasterServletModule</Gerrit-HttpModule>
 
               <Implementation-Vendor>Gerrit Code Review</Implementation-Vendor>
               <Implementation-URL>http://code.google.com/p/gerrit/</Implementation-URL>
@@ -70,6 +71,8 @@ limitations under the License.
         <directory>./</directory>
         <includes>
           <include>Documentation/**/*</include>
+          <include>static/**/*</include>
+          <include>html/**/*</include>
         </includes>
       </resource>
     </resources>
diff --git a/src/main/java/com/google/gerrit/plugins/multimaster/DataAggregator.java b/src/main/java/com/google/gerrit/plugins/multimaster/DataAggregator.java
new file mode 100644
index 0000000..0a39969
--- /dev/null
+++ b/src/main/java/com/google/gerrit/plugins/multimaster/DataAggregator.java
@@ -0,0 +1,120 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.plugins.multimaster;
+
+import java.util.Collection;
+import java.util.LinkedList;
+
+import com.google.gerrit.plugins.multimaster.cache.EvictNotice;
+import com.google.gerrit.plugins.multimaster.cache.Evictor;
+import com.google.gerrit.plugins.multimaster.peer.OutdatedThreshold;
+import com.google.gerrit.plugins.multimaster.peer.Peer;
+import com.google.gerrit.plugins.multimaster.peer.PeerActivity;
+import com.google.gerrit.plugins.multimaster.peer.PeerRegistry;
+import com.google.gerrit.plugins.multimaster.peer.Self;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+@Singleton
+public class DataAggregator implements Evictor.Listener {
+  public static class Status {
+    public Peer self;
+    public Long outdatedThreshold;
+    public Collection<PeerActivity> activities;
+    public Collection<Evict> recentEvicts;
+  }
+
+  public static class Evict {
+    public String pluginName;
+    public String cacheName;
+    public String key;
+
+    public Evict(String pluginName, String cacheName, String key) {
+      this.pluginName = pluginName;
+      this.cacheName = cacheName;
+      this.key = key;
+    }
+  }
+
+  private Peer self;
+  private long outdatedThreshold;
+  private PeerRegistry peerRegistry;
+
+  private LinkedList<Evict> recentEvicts = new LinkedList<Evict>();
+
+  @Inject
+  public DataAggregator(@Self Peer self, PeerRegistry peerRegistry,
+      @OutdatedThreshold Long threshold, Evictor evictor) {
+    this.self = self;
+    this.outdatedThreshold = threshold;
+    this.peerRegistry = peerRegistry;
+
+    evictor.addListener(this);
+  }
+
+  public String get() {
+    Status status = new Status();
+    status.self = this.self;
+    status.outdatedThreshold = this.outdatedThreshold;
+    status.activities = peerRegistry.getActivities();
+    status.recentEvicts = recentEvicts;
+
+    Gson gson = new Gson();
+
+    return gson.toJson(status);
+  }
+
+  @Override
+  public void onEvictNotice(final EvictNotice notice) {
+    recentEvicts.addFirst(new Evict(notice.pluginName, notice.cacheName,
+        notice.keyJson));
+
+    if (recentEvicts.size() > 20) {
+      recentEvicts.removeLast();
+    }
+  }
+
+  @Override
+  public void onOutdated() {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public void onAllOutdatedRestored() {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public void onCannotNotify(Peer target) {
+    // TODO Auto-generated method stub
+
+  }
+}
\ No newline at end of file
diff --git a/src/main/java/com/google/gerrit/plugins/multimaster/MultiMasterModule.java b/src/main/java/com/google/gerrit/plugins/multimaster/MultiMasterModule.java
index 380caad..46b0b8b 100644
--- a/src/main/java/com/google/gerrit/plugins/multimaster/MultiMasterModule.java
+++ b/src/main/java/com/google/gerrit/plugins/multimaster/MultiMasterModule.java
@@ -72,6 +72,8 @@ public class MultiMasterModule extends AbstractModule {
     DynamicSet.bind(binder(), CacheRemovalListener.class).to(
         CacheRemovalListenerImpl.class);
 
+    bind(DataAggregator.class).in(Scopes.SINGLETON);
+
     bind(StateMachine.class).in(Scopes.SINGLETON);
 
     bind(LifecycleListener.class).annotatedWith(UniqueAnnotations.create()).to(
diff --git a/src/main/java/com/google/gerrit/plugins/multimaster/servlet/JsonServlet.java b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/JsonServlet.java
new file mode 100644
index 0000000..edaddc5
--- /dev/null
+++ b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/JsonServlet.java
@@ -0,0 +1,58 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.plugins.multimaster.servlet;
+
+import java.io.IOException;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.gerrit.plugins.multimaster.DataAggregator;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+
+@Singleton
+public class JsonServlet extends HttpServlet {
+  private static final long serialVersionUID = 7211543843372263275L;
+
+  DataAggregator data;
+
+  @Inject
+  JsonServlet(DataAggregator data) {
+    this.data = data;
+  }
+
+  protected void doGet(HttpServletRequest req, HttpServletResponse res)
+      throws IOException {
+    res.setContentType("text/plain");
+    res.setCharacterEncoding("UTF-8");
+
+    res.getWriter().write(data.get());
+  }
+}
\ No newline at end of file
diff --git a/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServlet.java b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServlet.java
new file mode 100644
index 0000000..7abd7e7
--- /dev/null
+++ b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServlet.java
@@ -0,0 +1,78 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.plugins.multimaster.servlet;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.inject.Singleton;
+
+@Singleton
+public class MultiMasterServlet extends HttpServlet {
+  private static final long serialVersionUID = 7211543843372263275L;
+
+  protected void doGet(HttpServletRequest req, HttpServletResponse res)
+      throws IOException {
+    res.setContentType("text/html");
+    res.setCharacterEncoding("UTF-8");
+
+    res.getWriter().write(getContents());
+  }
+
+  private String getContents() {
+    InputStream stream =
+        this.getClass().getResourceAsStream("/html/index.html");
+
+    if (stream == null) {
+      return "Resource could not be found";
+    }
+
+    String contents = "";
+
+    try {
+      BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
+
+      String line;
+      while ((line = reader.readLine()) != null) {
+        contents += line;
+      }
+
+      reader.close();
+    } catch (IOException e) {
+      contents = e.getMessage();
+    }
+
+    return contents;
+  }
+}
\ No newline at end of file
diff --git a/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServletModule.java b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServletModule.java
new file mode 100644
index 0000000..490868a
--- /dev/null
+++ b/src/main/java/com/google/gerrit/plugins/multimaster/servlet/MultiMasterServletModule.java
@@ -0,0 +1,37 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.plugins.multimaster.servlet;
+
+import com.google.inject.servlet.ServletModule;
+
+public class MultiMasterServletModule extends ServletModule {
+  protected void configureServlets() {
+    serve("/status/").with(MultiMasterServlet.class);
+    serve("/status/json/").with(JsonServlet.class);
+  }
+}
\ No newline at end of file
diff --git a/static/css/status.css b/static/css/status.css
new file mode 100644
index 0000000..90428c8
--- /dev/null
+++ b/static/css/status.css
@@ -0,0 +1,42 @@
+/* Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+
+ Redistribution and use in source and binary forms, with or without
+ modification, are permitted provided that the following conditions are
+ met:
+ * Redistributions of source code must retain the above copyright
+ notice, this list of conditions and the following disclaimer.
+ * Redistributions in binary form must reproduce the above
+ copyright notice, this list of conditions and the following
+ disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of Code Aurora Forum, Inc. nor the names of its
+ contributors may be used to endorse or promote products derived
+ from this software without specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+ WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+ BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+ OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+ IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */
+
+body {
+	font-family: 'Source Sans Pro', sans-serif;
+}
+
+h1 {
+	font-size: 25px;
+}
+
+.peerStatustable th, td {
+	padding: 5px 8px;
+}
+
+.peerStatustable th {
+	background: #ECECEC;
+}
\ No newline at end of file
diff --git a/static/js/status.js b/static/js/status.js
new file mode 100644
index 0000000..4e54a92
--- /dev/null
+++ b/static/js/status.js
@@ -0,0 +1,98 @@
+// Copyright (c) 2012, Code Aurora Forum. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+// * Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following
+// disclaimer in the documentation and/or other materials provided
+// with the distribution.
+// * Neither the name of Code Aurora Forum, Inc. nor the names of its
+// contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+$(document).ready(function() {
+	makeStatusTable();
+	window.setInterval(function() {
+		makeStatusTable();
+	}, 5000);
+});
+
+function makeStatusTable() {
+	var peerStatusTable = $("<table class='peerStatusTable'></table>")
+	peerStatusTable
+			.html($("<tr><th>#</th><th>Id</th><th>Last Seen</th><th>Address</th><th>Port</th></tr>"));
+
+	var evictStatusTable = $("<table class='peerStatusTable'></table>")
+	evictStatusTable
+			.html($("<tr><th>#</th><th>Plugin</th><th>Cache</th><th>Key</th></tr>"));
+
+	$.getJSON('../status/json/', function(data) {
+		var outdated_threshold = data.outdatedThreshold;
+
+		var selfReceiveHost = $
+				.parseJSON(data.self.properties.receiveHost.string);
+		var selfReceivePort = $
+				.parseJSON(data.self.properties.receivePort.string);
+
+		peerStatusTable.append("<tr><td>Self</td><td>" + data.self.id.string
+				+ "</td><td>-</td><td>" + selfReceiveHost + "</td><td>"
+				+ selfReceivePort + "</td></tr>");
+
+		data.activities.sort(function(a, b) {
+			if (a.peer.id == b.peer.id) {
+				return 0;
+			}
+
+			return b.lastSeen - a.lastSeen;
+		});
+
+		$.each(data.activities, function(index, activity) {
+			var receiveHost = $
+					.parseJSON(activity.peer.properties.receiveHost.string);
+			var receivePort = $
+					.parseJSON(activity.peer.properties.receivePort.string);
+
+			var date = new Date(activity.lastSeen);
+
+			var status_row = $("<tr><td>" + index + "</td><td>"
+					+ activity.peer.id.string + "</td><td>" + date.toString()
+					+ "</td><td>" + receiveHost + "</td><td>" + receivePort
+					+ "</td></tr>");
+
+			if (Date.now() - activity.lastSeen < outdated_threshold) {
+				status_row.css('background-color', "#B6DB49");
+			} else {
+				status_row.css('background-color', "#FF7979");
+			}
+
+			peerStatusTable.append(status_row)
+		});
+
+		$.each(data.recentEvicts,
+				function(index, evict) {
+					var status_row = $("<tr><td>" + evict.pluginName
+							+ "</td><td>" + evict.cacheName + "</td><td>"
+							+ evict.key + "</td></tr>");
+
+					evictStatusTable.append(status_row)
+				});
+
+		$("#wrapper").html(peerStatusTable);
+		$("#wrapper").append(evictStatusTable);
+	});
+}
\ No newline at end of file
-- 
1.7.8

